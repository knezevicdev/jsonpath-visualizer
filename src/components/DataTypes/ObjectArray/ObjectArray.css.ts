import styled, { css } from 'styled-components';

interface WrapperProps {
  collapsed: boolean;
}

export const Wrapper = styled.div<WrapperProps>`
  margin-left: 20px;

  ${({ collapsed }) =>
    !collapsed &&
    css`
      display: flex;
    `}
`;

export const Name = styled.span`
  display: inline-block;
  color: green;
  letter-spacing: 0.5px;
  font-style: none;
  vertical-align: top;
  opacity: 0.8;

  &:hover {
    opacity: 1;
  }
`;

interface ValueWrapperProps {
  collapsed: boolean;
}

export const ValueWrapper = styled.span<ValueWrapperProps>`
  ${({ collapsed }) =>
    !collapsed &&
    css`
      display: flex;
    `}
`;

export const IconWrapper = styled.span`
  cursor: pointer;
`;

type ContentWrapperProps = {
  collapsed: boolean;
};

export const ContentWrapper = styled.div<ContentWrapperProps>`
  ${({ collapsed }) =>
    !collapsed &&
    css`
      display: none;
    `}
`;
