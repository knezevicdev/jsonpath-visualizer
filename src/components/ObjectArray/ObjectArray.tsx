import React, { FunctionComponent, useRef } from 'react';
import PropTypes from 'prop-types';

import { Variable, Collapse, ArrayGroup } from 'components';
import { ARRAY_GROUP_LIMIT } from 'utils/constants';
import { toType } from 'utils/helpers';

type ObjectProps = {
  name?: string;
  src: any;
  type: string;
  indexOffset?: number;
  opened?: boolean;
  path: string;
};

const ObjectArray: FunctionComponent<ObjectProps> = ({ src, type, name, indexOffset, opened, path }) => {
  const elements = useRef<JSX.Element[]>([]);

  const renderContent = () => {
    if (elements.current.length) return elements.current;

    const keys = Object.keys(src);

    keys.forEach((key) => {
      const variable = new JsonVariable(key, src[key], indexOffset, type, path);

      if (!src.hasOwnProperty(key)) {
        return;
      }

      if (variable.type === 'array' && (variable.value as any[])?.length > ARRAY_GROUP_LIMIT) {
        elements.current.push(
          <ArrayGroup key={variable.path} array={variable.value as any[]} name={variable.name} path={variable.path} />,
        );
      } else if (variable.type === 'object' || variable.type === 'array') {
        elements.current.push(
          <ObjectArray
            key={variable.path}
            name={variable.name}
            src={variable.value}
            type={variable.type}
            path={variable.path}
          />,
        );
      } else {
        elements.current.push(
          <Variable
            key={variable.path}
            name={variable.name}
            value={variable.value}
            type={variable.type}
            path={variable.path}
          />,
        );
      }
    });

    return elements.current;
  };

  return (
    <Collapse
      opened={!!opened}
      isArray={type === 'array'}
      name={name || ''}
      path={path}
      renderContent={renderContent}
    />
  );
};

ObjectArray.propTypes = {
  name: PropTypes.string.isRequired,
  src: PropTypes.any.isRequired,
  type: PropTypes.string.isRequired,
  indexOffset: PropTypes.number.isRequired,
  opened: PropTypes.bool,
  path: PropTypes.string.isRequired,
};

ObjectArray.defaultProps = {
  name: '',
  indexOffset: 0,
  opened: true,
};

class JsonVariable {
  public type: string;
  public path: string;

  constructor(public name: string, public value: unknown, indexOffset: number | undefined, type: string, path: string) {
    this.type = toType(value);
    if (type === 'array' && indexOffset) {
      this.name = String(Number(name || '0') + indexOffset);
    }

    if (path.length === 0) {
      this.path = this.name;
    } else {
      this.path = `${path}.${this.name}`;
    }
  }
}

export default ObjectArray;
