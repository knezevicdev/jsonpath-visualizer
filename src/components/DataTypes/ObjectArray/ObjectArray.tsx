import React, { FunctionComponent, useState, useRef } from 'react';
import PropTypes from 'prop-types';

import { toType } from 'utils/helpers';

import { CollapsedIcon, ExpandedIcon } from 'components/ToggleIcons/ToggleIcons';
import Ellipsis from 'components/Ellipsis/Ellipsis';
import Variable from 'components/Variable/Variable';

import { Wrapper, Name, IconWrapper, ContentWrapper, ValueWrapper } from './ObjectArray.css';
import { ARRAY_GROUP_LIMIT } from 'utils/constants';
import ArrayGroup from 'components/ArrayGroup/ArrayGroup';
import { observer } from 'mobx-react';
import { useStore } from 'utils/store';

type ObjectProps = {
  name?: string;
  src: any;
  type: string;
  indexOffset?: number;
  expanded?: boolean;
  pointer: string;
};

const ObjectArray: FunctionComponent<ObjectProps> = ({ src, type, name, indexOffset, expanded, pointer }) => {
  const [collapsed, setCollapsed] = useState(!!expanded);
  const elements = useRef<JSX.Element[]>([]);
  const store = useStore();

  const toggleCollapsed = () => {
    setCollapsed(!collapsed);
  };

  const renderContent = () => {
    if (elements.current.length) return elements.current;

    const keys = Object.keys(src);

    console.log(keys);

    keys.forEach((key) => {
      const variable = new JsonVariable(key, src[key], indexOffset, type, pointer);

      if (!src.hasOwnProperty(key)) {
        return;
      }

      if (variable.type === 'array' && (variable.value as any[])?.length > ARRAY_GROUP_LIMIT) {
        elements.current.push(
          <ArrayGroup
            key={variable.name}
            array={variable.value as any[]}
            name={variable.name}
            pointer={variable.pointer}
          />,
        );
      } else if (variable.type === 'object' || variable.type === 'array') {
        elements.current.push(
          <JSONObject
            key={variable.name}
            name={variable.name}
            src={variable.value}
            type={variable.type}
            pointer={variable.pointer}
          />,
        );
      } else {
        elements.current.push(
          <Variable
            key={variable.name}
            name={variable.name}
            value={variable.value}
            type={variable.type}
            pointer={variable.pointer}
          />,
        );
      }
    });

    return elements.current;
  };

  return (
    <Wrapper collapsed={collapsed}>
      <span>
        <IconWrapper onClick={toggleCollapsed}>{collapsed ? <ExpandedIcon /> : <CollapsedIcon />}</IconWrapper>{' '}
        {name?.length && <Name>{name}: </Name>}
      </span>
      <ValueWrapper collapsed={collapsed} data-matched={store.matched.includes(pointer)}>
        <span>{type === 'array' ? '[' : '{'}</span>
        {(elements.current.length || collapsed) && (
          <ContentWrapper collapsed={collapsed}>{renderContent()}</ContentWrapper>
        )}
        {!collapsed && (
          <div onClick={toggleCollapsed}>
            <Ellipsis />
          </div>
        )}
        <span>{type === 'array' ? ']' : '}'}</span>
      </ValueWrapper>
    </Wrapper>
  );
};

ObjectArray.propTypes = {
  name: PropTypes.string.isRequired,
  src: PropTypes.any.isRequired,
  type: PropTypes.string.isRequired,
  indexOffset: PropTypes.number.isRequired,
  expanded: PropTypes.bool,
  pointer: PropTypes.string.isRequired,
};

ObjectArray.defaultProps = {
  name: '',
  indexOffset: 0,
  expanded: true,
};

class JsonVariable {
  public type: string;
  public pointer: string;

  constructor(
    public name: string,
    public value: unknown,
    indexOffset: number | undefined,
    type: string,
    pointer: string,
  ) {
    this.type = toType(value);
    if (type === 'array' && indexOffset) {
      this.name = String(Number(name || '0') + indexOffset);
    }
    this.pointer = `${pointer}/${this.name}`;
  }
}

const JSONObject = observer(ObjectArray);

export default JSONObject;
