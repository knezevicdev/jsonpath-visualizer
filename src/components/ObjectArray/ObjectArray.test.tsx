import React from 'react';
import { shallow } from 'enzyme';

import ObjectArray from './ObjectArray';

describe('ObjectArray', () => {
  it('renders snapshots', () => {
    const wrapper = shallow(<ObjectArray opened={true} name="" path="" src={[]} type="array" />);
    expect(wrapper).toMatchSnapshot();
  });
});
