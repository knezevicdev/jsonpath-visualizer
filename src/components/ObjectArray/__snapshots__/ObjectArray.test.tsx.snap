// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ObjectArray renders snapshots 1`] = `
<Collapse
  isArray={true}
  name=""
  opened={true}
  path=""
  renderContent={[Function]}
/>
`;
